stage0: use cobra cli library

sgotti opened the pull request:
This PR converts rkt to use the cobra CLI library. Please see the single commits descriptions that explain the various choices and differences.

This should satisfy all the points from https://github.com/coreos/rkt/issues/632#issuecomment-108904070 except the bash/zsh completion (TODO).

Probably more functional tests to check that the various arguments are correctly handled should be useful.

jonboulle commented:
suuuuper cool, thanks sgotti!!

jonboulle commented:
@xiang90 pointed me towards https://github.com/spf13/pflag/pull/20 (and https://github.com/spf13/pflag/pull/18)

sgotti commented:
@jonboulle @xiang90 I started testing them. The two functionality looks like an improvement also if probably not blocking.
One avoids the need to provide an `=` between the long flag and its value.
The second one add the ability to specify any flag (not only a boolean) without providing a value.
This is different than the (common) case of not providing a flag (and using it's default value like done now for example in --local-config="/etc/rkt") (see the ls example in the upstream issue) and I don't know how much useful it will be in the rkt case.

jonboulle commented:
Yeah, the first sounds like something we want, not fussed about the second..

On Fri, Jun 12, 2015 at 7:31 AM, Simone Gotti <notifications@github.com>
wrote:

> @jonboulle <https://github.com/jonboulle> @xiang90
> <https://github.com/xiang90> I started testing them. The two
> functionality looks like an improvement also if probably not blocking.
> One avoids the need to provide an = between the long flag and its value.
> The second one add the ability to specify any flag (not only a boolean)
> without providing a value.
> This is different than the (common) case of not providing a flag (and
> using it's default value like done now for example in
> --local-config="/etc/rkt") (see the ls example in the upstream issue) and I
> don't know how much useful it will be in the rkt case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/rkt/pull/1028#issuecomment-111513246>.
>

eyakubovich commented:
Why can't we just have sane gnu style args (http://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html) with no extensions. Why is there this constant desire to make them more flexible? It just ends up being error prone to parse and understand. `</rant>`

eyakubovich commented:
LGTM other than nit above.

jonboulle commented:
This LGTM overall, I'm inclined to push it into 0.6.0.

@yifan-gu @alban @iaguis review/any concerns?

iaguis commented:
I didn't have the chance to look at it but it'd be cool to have in 0.6.0. I can review it tomorrow.

sgotti commented:
@eyakubovich Thanks for your review! Updated!

The TestPodManifest fails due to pflags not handling long arguments without "=" as @xiang90 noted. Should we wait for spf13/pflag#20 ?

@jonboulle What's missing is bash/zsh completion (I hope it will work as it's also used by kubernetes). I'd like to add also advanced argument completion (like image names, pod uuids) if it's possible.

What do you think about the custom help format? The default cobra doesn't have full deep command list but adds an additional "examples" section (that it's not actually implemented as no commands provides examples) and command aliases. They can be added to the help If there's a future need.

Another thing I'd like to heavily tests (in addition to the one used by the actual integration tests) is if some more complex commands continue to correctly work or have some problems caused by argument parsing.

iaguis commented on rkt/run.go:
```diff
@@ -27,29 +26,26 @@ import (
 	"strings"
 
 	"github.com/coreos/rkt/Godeps/_workspace/src/github.com/appc/spec/schema/types"
+	"github.com/coreos/rkt/Godeps/_workspace/src/github.com/spf13/cobra" // either set by linker, or guessed in init()
```
Shouldn't this comment be after `defaultStage1Image string`?

sgotti replied:
Uh! Dunno why how it ended there...

iaguis commented on rkt/enter.go:
```diff
@@ -17,43 +17,44 @@
-	enterFlags     flag.FlagSet
-	flagAppImageID types.Hash
+	cmdEnter = &cobra.Command{
+		Use:   "enter [--imageid IMAGEID] UUID [CMD [ARGS ...]]",
```
Since spf13/pflag#20 is not merged and for consistency we can try to write all the examples with the equals sign so they actually work.

```
Use:   "enter [--imageid=IMAGEID] UUID [CMD [ARGS ...]]"
```

sgotti replied:
Ok. Done for all the commands usage.

iaguis commented:
> The TestPodManifest fails due to pflags not handling long arguments without "=" as @xiang90 noted. Should we wait for spf13/pflag#20 ?

We can just fix the test:

```
diff --git a/tests/rkt_run_pod_manifest_test.go b/tests/rkt_run_pod_manifest_test.go
index 863d417..8bb0975 100644
--- a/tests/rkt_run_pod_manifest_test.go
+++ b/tests/rkt_run_pod_manifest_test.go
@@ -329,7 +329,7 @@ func TestPodManifest(t *testing.T) {
                defer os.Remove(manifestFile)
 
                // 1. Test 'rkt run'.
-               runCmd := fmt.Sprintf("%s run --pod-manifest %s", ctx.cmd(), manifestFile)
+               runCmd := fmt.Sprintf("%s run --pod-manifest=%s", ctx.cmd(), manifestFile)
                t.Logf("Running 'run' test #%v: %v", i, runCmd)
                child, err := gexpect.Spawn(runCmd)
                if err != nil {
@@ -351,7 +351,8 @@ func TestPodManifest(t *testing.T) {
                // 2. Test 'rkt prepare' + 'rkt run-prepared'.
                cmds := strings.Fields(ctx.cmd())
                prepareCmd := exec.Command(cmds[0], cmds[1:]...)
-               prepareCmd.Args = append(prepareCmd.Args, "--insecure-skip-verify", "prepare", "--pod-manifest", manifestFile)
+               prepareArg := fmt.Sprintf("--pod-manifest=%s", manifestFile)
+               prepareCmd.Args = append(prepareCmd.Args, "--insecure-skip-verify", "prepare", prepareArg)
                output, err := prepareCmd.Output()
                if err != nil {
                        t.Fatalf("Cannot read the output: %v", err)
```

I'd like to merge this even if `--flag ARG` doesn't work and then, when spf13/pflag#20 is merged we can bump it.

LGTM in general.

iaguis commented:
Can you update Documentation/app-container.md, Documentation/commands.md and README.md to use `=` too?

sgotti commented:
@iaguis thanks. I replaced all the occurrences that I found in the various files. IMHO sometimes this is quite ugly (like --set-env=ENV=VALUE)...

Also added the double dash to single dash long flags.

jonboulle commented on rkt/imageslist.go:
```diff
@@ -138,15 +146,16 @@ func (isa *ImagesSortAsc) String() string {
 }
 
+func (isa *ImagesSortAsc) Type() string {
+	return "imageSortAsc"
```
imagesSortAsc

sgotti replied:
Done! Thanks!

yifan-gu commented:
Sorry, kept distracted from reviewing this. Will do today.

jonboulle commented:
LGTM. Let's get @yifan-gu to do a final review then ship this thing!

yifan-gu commented:
LGTM except the question above

iaguis commented:
LGTM

Can you file issues to address the questions mentioned in https://github.com/sgotti/rkt/commit/e56e44655ebdbd72e5abee5115295fdf03c51351#commitcomment-11721334 (if needed)? Thanks!

sgotti commented:
@iaguis Thanks! Created #1059 and #1061.
